# ELVTR Data Science Main Project

## Deliverables

01 Git Repository

Include all project code with a README file containing a high-level project description.

Example README guide: [Make a README](link-to-readme-guide)

Report

* Methodology, approach, and model selection rationale.
* Advantages and limitations of the chosen model.
* Architecture of the final solution.
* Considerations on deployment and scalability of the solution - i.e., how will the model be used in BAU by the business?
* Estimated impact/ROI of the project.

The gitHub repository can be found here:

https://github.com/Wattysaid/dsif-git-main-project/tree/main

# Data Science in Finance: Lending Club Loan Analysis

## Project Overview

Lending Club has tasked us with preparing a loan application dataset for analysis and predictive modeling. 

The key tasks include data cleaning, exploratory data analysis, and building a predictive model for loan classification. An optional component involves building a real-time scoring application.

**Project Objectives**:
- Clean and preprocess the data.
- Perform exploratory data analysis (EDA) to gain insights.
- Develop a predictive model for loan application approval.
- (Optional) Build a real-time scoring application.

**Dataset Description**:
The dataset consists of loan application records, including various financial metrics and the application status. The data dictionary is provided for understanding the attributes.

**Dataset Path**:
- CSV: `data/1-raw/lending-club-2007-2020Q3/Loan_status_2007-2020Q3-100ksample.csv`
- Data Dictionary: `data/1-raw/lending-club-2007-2020Q3/LCDataDictionary.xlsx`

We'll import pandas initially to support our data manipulation and data import.

In [1]:
import pandas as pd

## Load Data Set

Let's start by loading our data. There are three files in total. We'll focus on loading the data file and the data dictionary.

In [ ]:
# Load the data xlsx file as a dataframe
df = pd.read_csv("c:\\Users\\kiera\\OneDrive\\Documents\\GitHub\\dsif-git-main-project\\elvtr_main_project\\data\\raw\\Loan_status_2007-2020Q3\\Loan_status_2007-2020Q3-100k-Full-Data.csv")

# Clean headers in the existing DataFrame 'df'
df.columns = df.columns.str.strip().str.lower().str.replace(" ", "_")

# Display cleaned headers
print("Cleaned headers:", df.columns.tolist())

df.shape

Cleaned headers: ['unnamed:_0.1', 'unnamed:_0', 'id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv', 'term', 'int_rate', 'installment', 'grade', 'sub_grade', 'emp_title', 'emp_length', 'home_ownership', 'annual_inc', 'verification_status', 'issue_d', 'loan_status', 'pymnt_plan', 'url', 'purpose', 'title', 'zip_code', 'addr_state', 'dti', 'delinq_2yrs', 'earliest_cr_line', 'fico_range_low', 'fico_range_high', 'inq_last_6mths', 'mths_since_last_delinq', 'mths_since_last_record', 'open_acc', 'pub_rec', 'revol_bal', 'revol_util', 'total_acc', 'initial_list_status', 'out_prncp', 'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee', 'recoveries', 'collection_recovery_fee', 'last_pymnt_d', 'last_pymnt_amnt', 'next_pymnt_d', 'last_credit_pull_d', 'last_fico_range_high', 'last_fico_range_low', 'collections_12_mths_ex_med', 'mths_since_last_major_derog', 'policy_code', 'application_type', 'annual_inc_joint', 'dti_joint', 'verification_status

(100000, 143)

In [ ]:
df.head()

unnamed:_0.1  unnamed:_0         id  loan_amnt  funded_amnt  \
0       1746494       25450  167338079       4000         4000   
1       2370717       81861   71016917      24000        24000   
2       2264870      397109   39589826       5000         5000   
3        595422       15492  134798709      24000        24000   
4        562657       90591  127097355      14000        14000   

   funded_amnt_inv        term int_rate  installment grade  ...  \
0           4000.0   36 months   13.08%       134.93     B  ...   
1          24000.0   60 months    9.16%       500.07     B  ...   
2           5000.0   36 months   10.49%       162.49     B  ...   
3          24000.0   60 months   11.05%       522.42     B  ...   
4          14000.0   60 months   13.59%       322.79     C  ...   

  hardship_start_date hardship_end_date payment_plan_start_date  \
0                 NaN               NaN                     NaN   
1                 NaN               NaN                     NaN   
2                 NaN               NaN                     NaN   
3                 NaN               NaN                     NaN   
4                 NaN               NaN                     NaN   

  hardship_length  hardship_dpd hardship_loan_status  \
0             NaN           NaN                  NaN   
1             NaN           NaN                  NaN   
2             NaN           NaN                  NaN   
3             NaN           NaN                  NaN   
4             NaN           NaN                  NaN   

  orig_projected_additional_accrued_interest hardship_payoff_balance_amount  \
0                                        NaN                            NaN   
1                                        NaN                            NaN   
2                                        NaN                            NaN   
3                                        NaN                            NaN   
4                                        NaN                            NaN   

  hardship_last_payment_amount debt_settlement_flag  
0                          NaN                    N  
1                          NaN                    N  
2                          NaN                    N  
3                          NaN                    N  
4                          NaN                    N  

[5 rows x 143 columns]

In [ ]:
# Load the data dictionary CSV file as a dataframe
df_data_dict = pd.read_excel("c:\\Users\\kiera\\OneDrive\\Documents\\GitHub\\dsif-git-main-project\\elvtr_main_project\\data\\raw\\Loan_status_2007-2020Q3\\LCDataDictionary.xlsx")

# Clean headers in the existing DataFrame 'df'
df_data_dict.columns = df_data_dict.columns.str.strip().str.lower().str.replace(" ", "_")

# Remove trailing whitespaces in all string columns of df_data_dict
df_data_dict = df_data_dict.apply(lambda x: x.str.strip() if x.dtype == "object" else x)

# Display cleaned headers
print("Cleaned headers:", df_data_dict.columns.tolist())

df_data_dict.shape

Cleaned headers: ['loanstatnew', 'description']


(153, 2)

In [ ]:
# Copies the columns and descriptions from the data dictionary into a data frame for future recall.
# Initialize empty lists for LoanStatNew and Description
loanstatnew = []
description = []

# Iterate through each row in the DataFrame and populate lists
for _, row in df_data_dict.iterrows():
    loanstatnew.append(row['loanstatnew'])
    description.append(row['description'])

# Apply left-aligned styling to both headers and data cells
styled_df_data_dict = df_data_dict.style.set_properties(
    **{'text-align': 'left', 'white-space': 'nowrap'}
).set_table_styles(
    [{'selector': 'th', 'props': [('text-align', 'left')]}]
)

# Display styled DataFrame
# styled_df_data_dict # comment out the displaying of the data for now to declutter 

After analysing our data dictionary it is possible to class our columns into figurative categories to better organise our analysis. 

These categories are, for now:

- Credit history, 
- Current Debt and Payment behaviours, 
- Employement, 
- Credit inquiries
- Loan Application information
- Hardship and Settlement Information
- Co-Borrower Information
- Loan Performance

Let's create a table for reference. We'll add these manually so that we can tweak the data within each group as we discover more about our data.

## Display Basic Data Discovery

Let's look at the information within our data frame (df) looking at our initial feature set (pre_hardship_fields).

In [ ]:
df.head()

unnamed:_0.1  unnamed:_0         id  loan_amnt  funded_amnt  \
0       1746494       25450  167338079       4000         4000   
1       2370717       81861   71016917      24000        24000   
2       2264870      397109   39589826       5000         5000   
3        595422       15492  134798709      24000        24000   
4        562657       90591  127097355      14000        14000   

   funded_amnt_inv        term int_rate  installment grade  ...  \
0           4000.0   36 months   13.08%       134.93     B  ...   
1          24000.0   60 months    9.16%       500.07     B  ...   
2           5000.0   36 months   10.49%       162.49     B  ...   
3          24000.0   60 months   11.05%       522.42     B  ...   
4          14000.0   60 months   13.59%       322.79     C  ...   

  hardship_start_date hardship_end_date payment_plan_start_date  \
0                 NaN               NaN                     NaN   
1                 NaN               NaN                     NaN   
2                 NaN               NaN                     NaN   
3                 NaN               NaN                     NaN   
4                 NaN               NaN                     NaN   

  hardship_length  hardship_dpd hardship_loan_status  \
0             NaN           NaN                  NaN   
1             NaN           NaN                  NaN   
2             NaN           NaN                  NaN   
3             NaN           NaN                  NaN   
4             NaN           NaN                  NaN   

  orig_projected_additional_accrued_interest hardship_payoff_balance_amount  \
0                                        NaN                            NaN   
1                                        NaN                            NaN   
2                                        NaN                            NaN   
3                                        NaN                            NaN   
4                                        NaN                            NaN   

  hardship_last_payment_amount debt_settlement_flag  
0                          NaN                    N  
1                          NaN                    N  
2                          NaN                    N  
3                          NaN                    N  
4                          NaN                    N  

[5 rows x 143 columns]

Our data contains 143 columns and 99999 rows of data. It is comprised of numerical (float, int) and categorical data (object)

We've ran into an error (KeyError: "['desc', 'member_id', 'verified_status_joint'] not in index"), for the time being we'll remove the missing fields from df_data_dict. These are most likely naming issues i.e. member_id is most likely id but considering we won't be using this data for now we can remove it and correct later on.

In [ ]:
# Extract column headers from df and the first column of df_data_dict
df_columns = set(df.columns)
df_data_dict_columns = set(df_data_dict.iloc[:, 0])  # First column of df_data_dict

# Find columns in df that are missing in df_data_dict
missing_in_data_dict = df_columns - df_data_dict_columns

# Find columns in df_data_dict that are missing in df
missing_in_df = df_data_dict_columns - df_columns

# Output the results
print("Columns in df that are missing in df_data_dict:")
print(missing_in_data_dict)

print("\nColumns in df_data_dict that are missing in df:")
print(missing_in_df)

Columns in df that are missing in df_data_dict:
{'unnamed:_0.1', 'unnamed:_0', 'verification_status_joint'}

Columns in df_data_dict that are missing in df:
{'settlement_date', 'debt_settlement_flag_date', 'member_id', 'settlement_percentage', 'verified_status_joint', 'settlement_amount', 'disbursement_method', 'settlement_status', 'desc', nan, 'sec_app_mths_since_last_major_derog', 'settlement_term'}


In [ ]:
# This list was created to run analysis later on. I've opted to select the pre_hardship_fields as my feature selection draft
# Subjective groupings created for post ML analysis. Due to the transformation these lists can only be used with df
groups = {
    "Credit History": [
        'earliest_cr_line', 'fico_range_high', 'fico_range_low', 'last_fico_range_high',
        'last_fico_range_low', 'mo_sin_old_il_acct', 'mo_sin_old_rev_tl_op', 'num_accts_ever_120_pd',
        'num_tl_120dpd_2m', 'pub_rec', 'pub_rec_bankruptcies'
    ],
    "Current Debt and Payment Behaviors": [
        'acc_now_delinq', 'all_util', 'bc_open_to_buy', 'bc_util', 'chargeoff_within_12_mths',
        'collections_12_mths_ex_med', 'delinq_2yrs', 'delinq_amnt', 'max_bal_bc', 'mths_since_last_delinq',
        'num_rev_accts', 'num_rev_tl_bal_gt_0', 'percent_bc_gt_75', 'revol_bal', 'revol_util',
        'tot_coll_amt', 'tot_cur_bal', 'tot_hi_cred_lim', 'total_bal_ex_mort', 'total_bc_limit'
    ],
    "Employment": [
        'emp_length', 'emp_title', 'annual_inc', 'annual_inc_joint'
    ],
    "Credit Inquiries": [
        'inq_fi', 'inq_last_12m', 'inq_last_6mths', 'num_tl_op_past_12m'
    ],
    "Loan Application Information": [
        'loan_amnt', 'term', 'int_rate', 'application_type', 'grade', 'sub_grade', 'purpose',
        'issue_d', 'home_ownership', 'zip_code', 'addr_state', 'title', 'desc', 'url'
    ],
    "Hardship and Settlement Information": [
        'hardship_flag', 'hardship_type', 'hardship_reason', 'hardship_status', 'hardship_start_date',
        'hardship_end_date', 'hardship_amount', 'hardship_length', 'settlement_status', 'settlement_date',
        'settlement_amount', 'settlement_percentage', 'settlement_term'
    ],
    "Co-Borrower Information": [
        'annual_inc_joint', 'dti_joint', 'verified_status_joint', 'revol_bal_joint',
        'sec_app_fico_range_low', 'sec_app_fico_range_high', 'sec_app_earliest_cr_line',
        'sec_app_mort_acc', 'sec_app_open_acc', 'sec_app_revol_util'
    ],
    "Loan Performance": [
        'funded_amnt', 'funded_amnt_inv', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
        'total_pymnt_inv', 'total_rec_int', 'total_rec_late_fee', 'total_rec_prncp', 'recoveries',
        'collection_recovery_fee', 'last_pymnt_amnt', 'last_pymnt_d', 'next_pymnt_d'
    ]
}

# Define pre and post hardship fields
pre_hardship_fields = [
    'acc_now_delinq', 'acc_open_past_24mths', 'addr_state', 'all_util', 'annual_inc', 
    'annual_inc_joint', 'application_type', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util', 
    'chargeoff_within_12_mths', 'collections_12_mths_ex_med', 'delinq_2yrs', 'delinq_amnt', 
    'dti', 'dti_joint', 'earliest_cr_line', 'emp_length', 'emp_title', 
    'fico_range_high', 'fico_range_low', 'funded_amnt', 'funded_amnt_inv', 'grade', 
    'home_ownership', 'il_util', 'initial_list_status', 'inq_fi', 'inq_last_12m', 
    'inq_last_6mths', 'installment', 'int_rate', 'issue_d', 'loan_amnt', 'loan_status', 
    'max_bal_bc', 'mo_sin_old_il_acct', 'mo_sin_old_rev_tl_op', 
    'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl', 'mort_acc', 'mths_since_last_delinq', 
    'mths_since_last_major_derog', 'mths_since_last_record', 'mths_since_rcnt_il', 
    'mths_since_recent_bc', 'mths_since_recent_bc_dlq', 'mths_since_recent_inq', 
    'mths_since_recent_revol_delinq', 'num_accts_ever_120_pd', 'num_actv_bc_tl', 
    'num_actv_rev_tl', 'num_bc_sats', 'num_bc_tl', 'num_il_tl', 'num_op_rev_tl', 
    'num_rev_accts', 'num_rev_tl_bal_gt_0', 'num_sats', 'num_tl_120dpd_2m', 'num_tl_30dpd', 
    'num_tl_90g_dpd_24m', 'num_tl_op_past_12m', 'open_acc', 'open_acc_6m', 'open_il_12m', 
    'open_il_24m', 'open_act_il', 'open_rv_12m', 'open_rv_24m', 'out_prncp', 
    'out_prncp_inv', 'pct_tl_nvr_dlq', 'percent_bc_gt_75', 'policy_code', 'pub_rec', 
    'pub_rec_bankruptcies', 'purpose', 'pymnt_plan', 'revol_bal', 'revol_util', 
    'sub_grade', 'tax_liens', 'term', 'title', 'tot_coll_amt', 'tot_cur_bal', 
    'tot_hi_cred_lim', 'total_acc', 'total_bal_ex_mort', 'total_bal_il', 'total_bc_limit', 
    'total_cu_tl', 'total_il_high_credit_limit', 'total_pymnt', 'total_pymnt_inv', 
    'total_rec_int', 'total_rec_late_fee', 'total_rec_prncp', 'total_rev_hi_lim', 
    'verification_status', 'zip_code'
]

post_hardship_fields = [
    'hardship_flag', 'hardship_type', 'hardship_reason', 'hardship_status', 
    'hardship_start_date', 'hardship_end_date', 'hardship_amount', 'hardship_length', 
    'hardship_dpd', 'hardship_loan_status', 'orig_projected_additional_accrued_interest', 
    'hardship_payoff_balance_amount', 'hardship_last_payment_amount', 'disbursement_method', 
    'debt_settlement_flag', 'debt_settlement_flag_date', 'settlement_status', 'settlement_date', 
    'settlement_amount', 'settlement_percentage', 'settlement_term'
]

In [ ]:
# Filter pre_hardship_fields to only include values that exist as columns in df
pre_hardship_fields_clean_kn = [field for field in pre_hardship_fields if field in df.columns]

# Display the filtered list
print("Filtered pre_hardship_fields:", pre_hardship_fields_clean_kn)

Filtered pre_hardship_fields: ['acc_now_delinq', 'acc_open_past_24mths', 'addr_state', 'all_util', 'annual_inc', 'annual_inc_joint', 'application_type', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util', 'chargeoff_within_12_mths', 'collections_12_mths_ex_med', 'delinq_2yrs', 'delinq_amnt', 'dti', 'dti_joint', 'earliest_cr_line', 'emp_length', 'emp_title', 'fico_range_high', 'fico_range_low', 'funded_amnt', 'funded_amnt_inv', 'grade', 'home_ownership', 'il_util', 'initial_list_status', 'inq_fi', 'inq_last_12m', 'inq_last_6mths', 'installment', 'int_rate', 'issue_d', 'loan_amnt', 'loan_status', 'max_bal_bc', 'mo_sin_old_il_acct', 'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl', 'mort_acc', 'mths_since_last_delinq', 'mths_since_last_major_derog', 'mths_since_last_record', 'mths_since_rcnt_il', 'mths_since_recent_bc', 'mths_since_recent_bc_dlq', 'mths_since_recent_inq', 'mths_since_recent_revol_delinq', 'num_accts_ever_120_pd', 'num_actv_bc_tl', 'num_actv_rev_tl', 'num_bc_sats'

In [ ]:
df[pre_hardship_fields_clean_kn].head()

acc_now_delinq  acc_open_past_24mths addr_state  all_util  annual_inc  \
0               0                   1.0         NY      59.0     48000.0   
1               0                   4.0         LA      67.0     57000.0   
2               0                   4.0         FL       NaN     55000.0   
3               0                   4.0         OH      25.0     43500.0   
4               0                   5.0         NE      43.0     48000.0   

   annual_inc_joint application_type  avg_cur_bal  bc_open_to_buy  bc_util  \
0               NaN       Individual      64756.0             NaN      NaN   
1               NaN       Individual       6182.0          7741.0     40.5   
2               NaN       Individual      22733.0          4521.0     54.8   
3               NaN       Individual       6973.0         43140.0     12.0   
4               NaN       Individual       5641.0         17318.0     39.9   

   ...  total_cu_tl  total_il_high_credit_limit   total_pymnt  \
0  ...          0.0                     17928.0    266.950000   
1  ...          7.0                     52954.0  25496.160000   
2  ...          NaN                      7526.0   5846.672836   
3  ...          0.0                     20000.0  12000.930000   
4  ...          0.0                     38181.0  16605.680000   

   total_pymnt_inv  total_rec_int  total_rec_late_fee total_rec_prncp  \
0           266.95          83.29                 0.0          183.66   
1         25496.16        5829.45                 0.0        19666.71   
2          5846.67         846.67                 0.0         5000.00   
3         12000.93        4318.62                 0.0         7682.31   
4         16605.68        2605.68                 0.0        14000.00   

  total_rev_hi_lim verification_status  zip_code  
0           2400.0     Source Verified     115xx  
1          35200.0     Source Verified     708xx  
2          11400.0        Not Verified     347xx  
3          54500.0        Not Verified     443xx  
4          43900.0     Source Verified     681xx  

[5 rows x 102 columns]

In [ ]:
df[pre_hardship_fields_clean_kn].describe()

acc_now_delinq  acc_open_past_24mths      all_util    annual_inc  \
count   100000.000000          98222.000000  70343.000000  1.000000e+05   
mean         0.003460              4.512166     56.154116  8.002918e+04   
std          0.063467              3.173066     20.804306  8.722434e+04   
min          0.000000              0.000000      0.000000  0.000000e+00   
25%          0.000000              2.000000     42.000000  4.716050e+04   
50%          0.000000              4.000000     57.000000  6.600000e+04   
75%          0.000000              6.000000     71.000000  9.500000e+04   
max          4.000000             42.000000    175.000000  1.099920e+07   

       annual_inc_joint    avg_cur_bal  bc_open_to_buy       bc_util  \
count      7.166000e+03   97560.000000    97152.000000  97101.000000   
mean       1.295764e+05   13757.234143    12779.260859     55.838138   
std        6.799755e+04   16693.219964    18099.386315     28.842058   
min        1.500000e+04       0.000000        0.000000      0.000000   
25%        8.603411e+04    3128.000000     2000.000000     32.700000   
50%        1.150000e+05    7461.500000     6262.500000     57.300000   
75%        1.550000e+05   19105.000000    16184.250000     81.200000   
max        1.110000e+06  471749.000000   393250.000000    175.000000   

       chargeoff_within_12_mths  collections_12_mths_ex_med  ...  \
count               99994.00000                99994.000000  ...   
mean                    0.00821                    0.017651  ...   
std                     0.10229                    0.144291  ...   
min                     0.00000                    0.000000  ...   
25%                     0.00000                    0.000000  ...   
50%                     0.00000                    0.000000  ...   
75%                     0.00000                    0.000000  ...   
max                     7.00000                    7.000000  ...   

        total_bal_il  total_bc_limit   total_cu_tl  \
count   70356.000000    98222.000000  70356.000000   
mean    36927.159276    24834.481664      1.517497   
std     46072.927224    24487.591805      2.719705   
min         0.000000        0.000000      0.000000   
25%      9120.000000     8900.000000      0.000000   
50%     24095.500000    17500.000000      0.000000   
75%     47534.750000    32500.000000      2.000000   
max    924163.000000   425150.000000     49.000000   

       total_il_high_credit_limit    total_pymnt  total_pymnt_inv  \
count                9.756200e+04  100000.000000    100000.000000   
mean                 4.561308e+04   12559.494957     12544.530188   
std                  4.693562e+04   10337.364917     10333.630505   
min                  0.000000e+00       0.000000         0.000000   
25%                  1.555025e+04    4744.560000      4732.010000   
50%                  3.408300e+04    9767.650000      9751.745000   
75%                  6.122675e+04   17567.367503     17540.212500   
max                  1.029490e+06   66775.524270     66775.520000   

       total_rec_int  total_rec_late_fee  total_rec_prncp  total_rev_hi_lim  
count  100000.000000       100000.000000    100000.000000      9.756200e+04  
mean     2479.307829            1.750509      9911.414064      3.639794e+04  
std      2766.019201           13.416653      8669.887647      3.582365e+04  
min         0.000000            0.000000         0.000000      0.000000e+00  
25%       724.002500            0.000000      3113.807500      1.540000e+04  
50%      1537.715000            0.000000      7415.485000      2.680000e+04  
75%      3175.680000            0.000000     14500.000000      4.570000e+04  
max     31714.370000          819.200000     40000.000000      1.333000e+06  

[8 rows x 83 columns]

In [ ]:
# Initialize list to collect cross-tab data
cross_tab_data = []

# Iterate over each group and each field to determine availability
for group, fields in groups.items():
    for field in fields:
        pre = 'Yes' if field in pre_hardship_fields else 'No'
        post = 'Yes' if field in post_hardship_fields else 'No'
        cross_tab_data.append([group, field, pre, post])

# Create DataFrame for cross-tab
cross_tab_df = pd.DataFrame(cross_tab_data, columns=["Group", "Field", "Pre-Hardship", "Post-Hardship"])

# Apply styling to left-align specific columns and set table header alignment
cross_tab_df_styled = cross_tab_df.style.set_properties(
    subset=['Group', 'Field'],
    **{'text-align': 'left'}
).set_table_styles(
    [{'selector': 'th.col_heading.level0', 'props': [('text-align', 'left')]}]
)

# Display the styled DataFrame
cross_tab_df_styled

Because are scoring is supposed to identify good and bad payers we'll leverage Pre-Hardship feature list for further analysis.

In [ ]:
#load the employtment mapping CSV file as a dataframe
df_emp_title = pd.read_csv("c:\\Users\\kiera\\OneDrive\\Documents\\GitHub\\dsif-git-main-project\\elvtr_main_project\\data\\raw\\emp_title_mapping.csv")

# Clean headers in the existing DataFrame
df_emp_title.columns = df_emp_title.columns.str.strip().str.lower().str.replace(" ", "_")

# Display cleaned headers
print("Cleaned headers:", df_emp_title.columns.tolist())

df_emp_title.shape

Cleaned headers: ['job_title', 'category']


(91, 2)

In [ ]:
# Initialize empty lists for LoanStatNew and Description
jobtitle = []
job_category = []

# Iterate through each row in the DataFrame and populate lists
for _, row in df_emp_title.iterrows():
    jobtitle.append(row['job_title'])
    job_category.append(row['category'])

# Apply left-aligned styling to both headers and data cells
styled_df_emp_title = df_emp_title.style.set_properties(
    **{'text-align': 'left', 'white-space': 'nowrap'}
).set_table_styles(
    [{'selector': 'th', 'props': [('text-align', 'left')]}]
)

# Display styled DataFrame
styled_df_emp_title

In [ ]:
# Cross table on 'category' in df_emp_title
category_crosstab = pd.crosstab(index=df_emp_title['category'], columns='count').sort_values(by='count', ascending=False)

# Display the crosstab
category_crosstab

col_0               count
category                 
Healthcare             10
Executive               7
IT/Technology           6
Education               6
Other                   5
Operations              5
Finance                 5
Administrative          5
Technical Support       4
Manufacturing           4
Security                3
Management              3
Customer Service        3
Sales                   2
Insurance               2
Quality Assurance       2
Logistics               2
Engineering             2
Procurement             2
Content/Media           1
Animal Care             1
Event Management        1
Religious               1
Project Management      1
Marketing               1
Construction            1
Consulting              1
Labor                   1
Agriculture             1
Human Resources         1
Hospitality             1
Emergency Services      1

Let's import our libraries and configure any paramaters for charting later on.

In [ ]:
# Essential libraries for data manipulation, statistics, and visualization
import numpy as np
import pandas as pd
from scipy import stats
from scipy.stats import normaltest, shapiro, anderson, kstest, skew

# Visualization libraries
import matplotlib.pyplot as plt  # For standard plotting
import seaborn as sns  # For static plots with themes
import plotly.express as px  # For interactive plots
import missingno as msno  # For missing data visualization

# Machine Learning libraries
from sklearn.preprocessing import StandardScaler, LabelEncoder, label_binarize
from sklearn.model_selection import train_test_split, cross_val_score  # Data splitting and cross-validation
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor
from sklearn.ensemble import (
    RandomForestClassifier, RandomForestRegressor,
    GradientBoostingClassifier, GradientBoostingRegressor
)
from sklearn.svm import SVC, SVR  # Support Vector Machines for classification and regression
from sklearn.neighbors import KNeighborsClassifier, KNeighborsRegressor
from sklearn.naive_bayes import GaussianNB  # Naive Bayes Classifier
from sklearn.cluster import KMeans  # K-Means clustering
from sklearn.decomposition import PCA  # Dimensionality reduction
from sklearn.multiclass import OneVsOneClassifier, OneVsRestClassifier
from sklearn.feature_selection import RFE  # Recursive Feature Elimination

# Additional machine learning models
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from catboost import CatBoostClassifier
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# Evaluation metrics
from sklearn.metrics import (
    accuracy_score, precision_score, recall_score, f1_score,
    roc_auc_score, confusion_matrix, roc_curve, precision_recall_curve, average_precision_score
)

# Utility libraries
from tqdm import tqdm
import joblib

# Pandas display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# Plot settings for consistent figure size (A4 landscape top half)
FIG_WIDTH = 11.69  # Width
FIG_HEIGHT = 4.14  # Height

# Set the theme for Seaborn plots
sns.set_theme(style='whitegrid')

c:\Users\kiera\AppData\Local\Programs\Python\Python312\Lib\site-packages\dask\dataframe\__init__.py:42: FutureWarning: 
Dask dataframe query planning is disabled because dask-expr is not installed.

You can install it with `pip install dask[dataframe]` or `conda install dask`.
This will raise in a future version.

  warnings.warn(msg, FutureWarning)


### Basic Data Overview

Before performing any analysis, we will explore the structure of the dataset to understand the nature of the available data. This includes checking the number of rows and columns, the data types of each feature, and identifying any missing values. Understanding these characteristics is essential for guiding data cleaning and feature engineering steps later in the analysis.

In [ ]:
df[pre_hardship_fields].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Columns: 102 entries, acc_now_delinq to zip_code
dtypes: float64(70), int64(13), object(19)
memory usage: 77.8+ MB
